File: erubit.platform/src/main/java/a/erubit/platform/course/PhraseLesson.java
package a.erubit.platform.course;

import android.support.annotation.NonNull;

import java.util.Random;

import u.U;

public class PhraseLesson extends BunchLesson {
    private static final int RANK_FAMILIAR = 1;
    private static final int RANK_LEARNED = 2;
    private static final int RANK_LEARNED_WELL = 3;

    private final Random mRandom;
    private final int mApxSize;

    PhraseLesson(Course course) {
        super(course);

        mRandom = new Random();
        mApxSize = 3 + mRandom.nextInt(3);
    }

    @Override
    protected int getRankFamiliar() {
        return RANK_FAMILIAR;
    }
    @Override
    protected int getRankLearned() {
        return RANK_LEARNED;
    }
    @Override
    protected int getRankLearnedWell() {
        return RANK_LEARNED_WELL;
    }


    @NonNull
    @Override
    protected PresentableDescriptor getPresentable(Item problemItem) {
        Problem problem = new Problem(this, problemItem);
        String[] words = U.defurigana(problem.meaning).split("\\s+");
        problem.variants = new String[words.length + mApxSize];
        System.arraycopy(words, 0, problem.variants, 0, words.length);
        for (int k = 0; k < mApxSize; k++)
            problem.variants[words.length + k] = mVariants.get(mRandom.nextInt(mVariants.size()));

        U.shuffleStrArray(problem.variants);

        return new PresentableDescriptor(problem);
    }
}

File: erubit.platform/src/main/java/a/erubit/platform/course/lesson/PhraseLesson.kt
package a.erubit.platform.course.lesson

import a.erubit.platform.course.Course
import u.U
import java.util.*


class PhraseLesson internal constructor(course: Course) : CharacterLesson(course) {
	private val mRandom: Random = Random()
	private val mApxSize: Int = 3 + mRandom.nextInt(3)

	override val rankFamiliar: Int
		get() = 1
	override val rankLearned: Int
		get() = 2
	override val rankLearnedWell: Int
		get() = 3

	override fun getPresentable(problemItem: BunchLesson.Item): PresentableDescriptor {
		if (problemItem !is Item)
			return PresentableDescriptor.ERROR

		val problem = Problem(this, problemItem)
		val words = U.defurigana(problem.meaning).split("\\s+").toTypedArray()

		problem.variants = arrayOfNulls(words.size + mApxSize)
		System.arraycopy(words, 0, problem.variants, 0, words.size)

		val variants = mNoise ?: return PresentableDescriptor.ERROR

		for (k in 0 until mApxSize)
			problem.variants[words.size + k] = variants[mRandom.nextInt(variants.size)]

		U.shuffleStrArray(problem.variants)

		return PresentableDescriptor(problem)
	}
}
